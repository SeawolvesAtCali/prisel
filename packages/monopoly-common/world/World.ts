import { serialize, update } from 'serializr';
import { genId } from '../genId';
import { GameObject, GameObjectClass } from './GameObject';
import { Id } from './Id';
import { Ref as Ref2 } from './ref2';

export interface SerializedWorld {
    [key: string]: Array<any>;
}

/**
 * World holds all game objects. Provide easy access to game object through ids
 */
export class World {
    private objectMap: Map<Id, GameObject> = new Map();
    private idMap: Map<string, Set<Id>> = new Map();

    constructor() {
        this.makeRef = this.makeRef.bind(this);
    }

    private createId<T extends GameObject>(specifiedId?: Id<T>): Id<T> {
        if (specifiedId && this.objectMap.has(specifiedId)) {
            throw new Error(
                `Cannot create game object of id ${specifiedId} because there is already one existing`,
            );
        }
        if (specifiedId) {
            return specifiedId;
        }
        return genId();
    }

    public get<T extends GameObject>(id: Id<T>): T | null {
        const object = this.objectMap.get(id);
        return (object as T) || null;
    }

    private getIdSetOfSameType(object: GameObject) {
        const type = object.type;
        // const gameObjectClass = this.gameObjectTypeRegistry.get(type);
        // if (gameObjectClass && object instanceof gameObjectClass) {
        const set = this.idMap.get(type);
        if (!set) {
            const newSet = new Set<string>();
            this.idMap.set(type, newSet);
            return newSet;
        }
        return set;
    }

    private getAllFromSet<T extends GameObject>(idSet: Set<Id>): T[] {
        return Array.from(idSet)
            .map((id) => this.objectMap.get(id) as T)
            .filter(Boolean);
    }

    public getAll<T extends GameObjectClass<any>>(typeOrClazz: string | T): InstanceType<T>[] {
        const typeString = typeof typeOrClazz === 'string' ? typeOrClazz : typeOrClazz.TYPE;
        const set = this.idMap.get(typeString);
        if (set) {
            return this.getAllFromSet<InstanceType<T>>(set);
        }
        return [];
    }

    /** Create a GameObject of given type, and add to world */
    public create<T extends GameObjectClass<any>>(
        clazz: T,
        id?: Id<InstanceType<T>>,
    ): InstanceType<T> {
        const object = new clazz();
        object.world = this;
        const objectId = this.createId<InstanceType<T>>(id);
        object.id = objectId;
        this.objectMap.set(objectId, object);
        this.getIdSetOfSameType(object).add(objectId);

        return object;
    }

    public createRef<T extends GameObjectClass<any>>(
        clazz: T,
        id?: Id<InstanceType<T>>,
    ): Ref2<InstanceType<T>> {
        return this.makeRef(this.create(clazz, id));
    }

    public add<T extends GameObject>(object: T) {
        if (this.idMap.get(object.type)?.has(object.id)) {
            throw new Error(`object ${object.type} already has an id ${object.id}`);
        }
        this.objectMap.set(object.id, object);
        this.getIdSetOfSameType(object).add(object.id);
        object.world = this;
    }

    public remove<T extends GameObject>(idOrGameObject: Id | T) {
        let deletedObject;
        if (typeof idOrGameObject === 'string') {
            deletedObject = this.objectMap.get(idOrGameObject);
            this.objectMap.delete(idOrGameObject);
        } else {
            deletedObject = idOrGameObject;
            this.objectMap.delete(idOrGameObject.id);
        }
        if (deletedObject) {
            this.getIdSetOfSameType(deletedObject).delete(deletedObject.id);
        }
    }

    public makeRef<T extends GameObject>(keyOrObject: Id<T> | T): Ref2<T> {
        return Ref2.of<T>(keyOrObject, this);
    }

    /**
     * Called before serializing. This method should be used to populate the
     * serializable fields in sub class.
     */
    protected populateSerializedFields(): void {}
    public serialize(): any {
        this.populateSerializedFields();
        const serialzied = serialize(this);
        this.clearSerializedFields();
        return serialzied;
    }

    /**
     * Called after deserialized. This method should be used to clear out the
     * serialized fields to reduce duplicate reference. All GameObjects should
     * already be stored inside the internal map of World.
     */
    protected clearSerializedFields(): void {}
    /**
     * deserialize into current world make sure this is an empty world,
     * otherwise we might have id collision
     * @param serialized
     */
    public populate(serialized: any) {
        this.clearSerializedFields();
        update(
            this,
            serialized,
            (err) => {
                if (err) {
                    console.error(err);
                } else {
                    this.clearSerializedFields();
                }
            },
            { world: this },
        );
        return this;
    }
}
