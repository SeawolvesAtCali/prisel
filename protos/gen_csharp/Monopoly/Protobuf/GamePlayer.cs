// <auto-generated>
//     Generated by the protocol buffer compiler.  DO NOT EDIT!
//     source: monopoly/game_player.proto
// </auto-generated>
#pragma warning disable 1591, 0612, 3021
#region Designer generated code

using pb = global::Google.Protobuf;
using pbc = global::Google.Protobuf.Collections;
using pbr = global::Google.Protobuf.Reflection;
using scg = global::System.Collections.Generic;
namespace Monopoly.Protobuf {

  /// <summary>Holder for reflection information generated from monopoly/game_player.proto</summary>
  public static partial class GamePlayerReflection {

    #region Descriptor
    /// <summary>File descriptor for monopoly/game_player.proto</summary>
    public static pbr::FileDescriptor Descriptor {
      get { return descriptor; }
    }
    private static pbr::FileDescriptor descriptor;

    static GamePlayerReflection() {
      byte[] descriptorData = global::System.Convert.FromBase64String(
          string.Concat(
            "Chptb25vcG9seS9nYW1lX3BsYXllci5wcm90bxIIbW9ub3BvbHkaGW1vbm9w",
            "b2x5L2Nvb3JkaW5hdGUucHJvdG8aGHByaXNlbC9wbGF5ZXJfaW5mby5wcm90",
            "byKHAQoKR2FtZVBsYXllchIKCgJpZBgBIAEoCRINCgVtb25leRgCIAEoBRIh",
            "CgNwb3MYAyABKAsyFC5tb25vcG9seS5Db29yZGluYXRlEhEKCWNoYXJhY3Rl",
            "chgEIAEoBRIoCgxib3VuZF9wbGF5ZXIYBSABKAsyEi5wcmlzZWwuUGxheWVy",
            "SW5mb0IUqgIRTW9ub3BvbHkuUHJvdG9idWZiBnByb3RvMw=="));
      descriptor = pbr::FileDescriptor.FromGeneratedCode(descriptorData,
          new pbr::FileDescriptor[] { global::Monopoly.Protobuf.CoordinateReflection.Descriptor, global::Prisel.Protobuf.PlayerInfoReflection.Descriptor, },
          new pbr::GeneratedClrTypeInfo(null, null, new pbr::GeneratedClrTypeInfo[] {
            new pbr::GeneratedClrTypeInfo(typeof(global::Monopoly.Protobuf.GamePlayer), global::Monopoly.Protobuf.GamePlayer.Parser, new[]{ "Id", "Money", "Pos", "Character", "BoundPlayer" }, null, null, null, null)
          }));
    }
    #endregion

  }
  #region Messages
  public sealed partial class GamePlayer : pb::IMessage<GamePlayer>
  #if !GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE
      , pb::IBufferMessage
  #endif
  {
    private static readonly pb::MessageParser<GamePlayer> _parser = new pb::MessageParser<GamePlayer>(() => new GamePlayer());
    private pb::UnknownFieldSet _unknownFields;
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public static pb::MessageParser<GamePlayer> Parser { get { return _parser; } }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public static pbr::MessageDescriptor Descriptor {
      get { return global::Monopoly.Protobuf.GamePlayerReflection.Descriptor.MessageTypes[0]; }
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    pbr::MessageDescriptor pb::IMessage.Descriptor {
      get { return Descriptor; }
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public GamePlayer() {
      OnConstruction();
    }

    partial void OnConstruction();

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public GamePlayer(GamePlayer other) : this() {
      id_ = other.id_;
      money_ = other.money_;
      pos_ = other.pos_ != null ? other.pos_.Clone() : null;
      character_ = other.character_;
      boundPlayer_ = other.boundPlayer_ != null ? other.boundPlayer_.Clone() : null;
      _unknownFields = pb::UnknownFieldSet.Clone(other._unknownFields);
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public GamePlayer Clone() {
      return new GamePlayer(this);
    }

    /// <summary>Field number for the "id" field.</summary>
    public const int IdFieldNumber = 1;
    private string id_ = "";
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public string Id {
      get { return id_; }
      set {
        id_ = pb::ProtoPreconditions.CheckNotNull(value, "value");
      }
    }

    /// <summary>Field number for the "money" field.</summary>
    public const int MoneyFieldNumber = 2;
    private int money_;
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public int Money {
      get { return money_; }
      set {
        money_ = value;
      }
    }

    /// <summary>Field number for the "pos" field.</summary>
    public const int PosFieldNumber = 3;
    private global::Monopoly.Protobuf.Coordinate pos_;
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public global::Monopoly.Protobuf.Coordinate Pos {
      get { return pos_; }
      set {
        pos_ = value;
      }
    }

    /// <summary>Field number for the "character" field.</summary>
    public const int CharacterFieldNumber = 4;
    private int character_;
    /// <summary>
    /// character determine the sprite set, or color client side use to
    /// denote the player
    /// </summary>
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public int Character {
      get { return character_; }
      set {
        character_ = value;
      }
    }

    /// <summary>Field number for the "bound_player" field.</summary>
    public const int BoundPlayerFieldNumber = 5;
    private global::Prisel.Protobuf.PlayerInfo boundPlayer_;
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public global::Prisel.Protobuf.PlayerInfo BoundPlayer {
      get { return boundPlayer_; }
      set {
        boundPlayer_ = value;
      }
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public override bool Equals(object other) {
      return Equals(other as GamePlayer);
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public bool Equals(GamePlayer other) {
      if (ReferenceEquals(other, null)) {
        return false;
      }
      if (ReferenceEquals(other, this)) {
        return true;
      }
      if (Id != other.Id) return false;
      if (Money != other.Money) return false;
      if (!object.Equals(Pos, other.Pos)) return false;
      if (Character != other.Character) return false;
      if (!object.Equals(BoundPlayer, other.BoundPlayer)) return false;
      return Equals(_unknownFields, other._unknownFields);
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public override int GetHashCode() {
      int hash = 1;
      if (Id.Length != 0) hash ^= Id.GetHashCode();
      if (Money != 0) hash ^= Money.GetHashCode();
      if (pos_ != null) hash ^= Pos.GetHashCode();
      if (Character != 0) hash ^= Character.GetHashCode();
      if (boundPlayer_ != null) hash ^= BoundPlayer.GetHashCode();
      if (_unknownFields != null) {
        hash ^= _unknownFields.GetHashCode();
      }
      return hash;
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public override string ToString() {
      return pb::JsonFormatter.ToDiagnosticString(this);
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public void WriteTo(pb::CodedOutputStream output) {
    #if !GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE
      output.WriteRawMessage(this);
    #else
      if (Id.Length != 0) {
        output.WriteRawTag(10);
        output.WriteString(Id);
      }
      if (Money != 0) {
        output.WriteRawTag(16);
        output.WriteInt32(Money);
      }
      if (pos_ != null) {
        output.WriteRawTag(26);
        output.WriteMessage(Pos);
      }
      if (Character != 0) {
        output.WriteRawTag(32);
        output.WriteInt32(Character);
      }
      if (boundPlayer_ != null) {
        output.WriteRawTag(42);
        output.WriteMessage(BoundPlayer);
      }
      if (_unknownFields != null) {
        _unknownFields.WriteTo(output);
      }
    #endif
    }

    #if !GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    void pb::IBufferMessage.InternalWriteTo(ref pb::WriteContext output) {
      if (Id.Length != 0) {
        output.WriteRawTag(10);
        output.WriteString(Id);
      }
      if (Money != 0) {
        output.WriteRawTag(16);
        output.WriteInt32(Money);
      }
      if (pos_ != null) {
        output.WriteRawTag(26);
        output.WriteMessage(Pos);
      }
      if (Character != 0) {
        output.WriteRawTag(32);
        output.WriteInt32(Character);
      }
      if (boundPlayer_ != null) {
        output.WriteRawTag(42);
        output.WriteMessage(BoundPlayer);
      }
      if (_unknownFields != null) {
        _unknownFields.WriteTo(ref output);
      }
    }
    #endif

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public int CalculateSize() {
      int size = 0;
      if (Id.Length != 0) {
        size += 1 + pb::CodedOutputStream.ComputeStringSize(Id);
      }
      if (Money != 0) {
        size += 1 + pb::CodedOutputStream.ComputeInt32Size(Money);
      }
      if (pos_ != null) {
        size += 1 + pb::CodedOutputStream.ComputeMessageSize(Pos);
      }
      if (Character != 0) {
        size += 1 + pb::CodedOutputStream.ComputeInt32Size(Character);
      }
      if (boundPlayer_ != null) {
        size += 1 + pb::CodedOutputStream.ComputeMessageSize(BoundPlayer);
      }
      if (_unknownFields != null) {
        size += _unknownFields.CalculateSize();
      }
      return size;
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public void MergeFrom(GamePlayer other) {
      if (other == null) {
        return;
      }
      if (other.Id.Length != 0) {
        Id = other.Id;
      }
      if (other.Money != 0) {
        Money = other.Money;
      }
      if (other.pos_ != null) {
        if (pos_ == null) {
          Pos = new global::Monopoly.Protobuf.Coordinate();
        }
        Pos.MergeFrom(other.Pos);
      }
      if (other.Character != 0) {
        Character = other.Character;
      }
      if (other.boundPlayer_ != null) {
        if (boundPlayer_ == null) {
          BoundPlayer = new global::Prisel.Protobuf.PlayerInfo();
        }
        BoundPlayer.MergeFrom(other.BoundPlayer);
      }
      _unknownFields = pb::UnknownFieldSet.MergeFrom(_unknownFields, other._unknownFields);
    }

    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    public void MergeFrom(pb::CodedInputStream input) {
    #if !GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE
      input.ReadRawMessage(this);
    #else
      uint tag;
      while ((tag = input.ReadTag()) != 0) {
        switch(tag) {
          default:
            _unknownFields = pb::UnknownFieldSet.MergeFieldFrom(_unknownFields, input);
            break;
          case 10: {
            Id = input.ReadString();
            break;
          }
          case 16: {
            Money = input.ReadInt32();
            break;
          }
          case 26: {
            if (pos_ == null) {
              Pos = new global::Monopoly.Protobuf.Coordinate();
            }
            input.ReadMessage(Pos);
            break;
          }
          case 32: {
            Character = input.ReadInt32();
            break;
          }
          case 42: {
            if (boundPlayer_ == null) {
              BoundPlayer = new global::Prisel.Protobuf.PlayerInfo();
            }
            input.ReadMessage(BoundPlayer);
            break;
          }
        }
      }
    #endif
    }

    #if !GOOGLE_PROTOBUF_REFSTRUCT_COMPATIBILITY_MODE
    [global::System.Diagnostics.DebuggerNonUserCodeAttribute]
    void pb::IBufferMessage.InternalMergeFrom(ref pb::ParseContext input) {
      uint tag;
      while ((tag = input.ReadTag()) != 0) {
        switch(tag) {
          default:
            _unknownFields = pb::UnknownFieldSet.MergeFieldFrom(_unknownFields, ref input);
            break;
          case 10: {
            Id = input.ReadString();
            break;
          }
          case 16: {
            Money = input.ReadInt32();
            break;
          }
          case 26: {
            if (pos_ == null) {
              Pos = new global::Monopoly.Protobuf.Coordinate();
            }
            input.ReadMessage(Pos);
            break;
          }
          case 32: {
            Character = input.ReadInt32();
            break;
          }
          case 42: {
            if (boundPlayer_ == null) {
              BoundPlayer = new global::Prisel.Protobuf.PlayerInfo();
            }
            input.ReadMessage(BoundPlayer);
            break;
          }
        }
      }
    }
    #endif

  }

  #endregion

}

#endregion Designer generated code
