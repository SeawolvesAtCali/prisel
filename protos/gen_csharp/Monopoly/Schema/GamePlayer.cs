// <auto-generated>
//  automatically generated by the FlatBuffers compiler, do not modify
// </auto-generated>

namespace Monopoly.Schema
{

using global::System;
using global::System.Collections.Generic;
using global::FlatBuffers;

public struct GamePlayer : IFlatbufferObject
{
  private Table __p;
  public ByteBuffer ByteBuffer { get { return __p.bb; } }
  public static void ValidateVersion() { FlatBufferConstants.FLATBUFFERS_2_0_0(); }
  public static GamePlayer GetRootAsGamePlayer(ByteBuffer _bb) { return GetRootAsGamePlayer(_bb, new GamePlayer()); }
  public static GamePlayer GetRootAsGamePlayer(ByteBuffer _bb, GamePlayer obj) { return (obj.__assign(_bb.GetInt(_bb.Position) + _bb.Position, _bb)); }
  public void __init(int _i, ByteBuffer _bb) { __p = new Table(_i, _bb); }
  public GamePlayer __assign(int _i, ByteBuffer _bb) { __init(_i, _bb); return this; }

  public string Id { get { int o = __p.__offset(4); return o != 0 ? __p.__string(o + __p.bb_pos) : null; } }
#if ENABLE_SPAN_T
  public Span<byte> GetIdBytes() { return __p.__vector_as_span<byte>(4, 1); }
#else
  public ArraySegment<byte>? GetIdBytes() { return __p.__vector_as_arraysegment(4); }
#endif
  public byte[] GetIdArray() { return __p.__vector_as_array<byte>(4); }
  public int Money { get { int o = __p.__offset(6); return o != 0 ? __p.bb.GetInt(o + __p.bb_pos) : (int)0; } }
  public Monopoly.Schema.Coordinate? Pos { get { int o = __p.__offset(8); return o != 0 ? (Monopoly.Schema.Coordinate?)(new Monopoly.Schema.Coordinate()).__assign(__p.__indirect(o + __p.bb_pos), __p.bb) : null; } }
  public int Character { get { int o = __p.__offset(10); return o != 0 ? __p.bb.GetInt(o + __p.bb_pos) : (int)0; } }
  public Prisel.Common.PlayerInfo? BoundPlayer { get { int o = __p.__offset(12); return o != 0 ? (Prisel.Common.PlayerInfo?)(new Prisel.Common.PlayerInfo()).__assign(__p.__indirect(o + __p.bb_pos), __p.bb) : null; } }

  public static Offset<Monopoly.Schema.GamePlayer> CreateGamePlayer(FlatBufferBuilder builder,
      StringOffset idOffset = default(StringOffset),
      int money = 0,
      Offset<Monopoly.Schema.Coordinate> posOffset = default(Offset<Monopoly.Schema.Coordinate>),
      int character = 0,
      Offset<Prisel.Common.PlayerInfo> bound_playerOffset = default(Offset<Prisel.Common.PlayerInfo>)) {
    builder.StartTable(5);
    GamePlayer.AddBoundPlayer(builder, bound_playerOffset);
    GamePlayer.AddCharacter(builder, character);
    GamePlayer.AddPos(builder, posOffset);
    GamePlayer.AddMoney(builder, money);
    GamePlayer.AddId(builder, idOffset);
    return GamePlayer.EndGamePlayer(builder);
  }

  public static void StartGamePlayer(FlatBufferBuilder builder) { builder.StartTable(5); }
  public static void AddId(FlatBufferBuilder builder, StringOffset idOffset) { builder.AddOffset(0, idOffset.Value, 0); }
  public static void AddMoney(FlatBufferBuilder builder, int money) { builder.AddInt(1, money, 0); }
  public static void AddPos(FlatBufferBuilder builder, Offset<Monopoly.Schema.Coordinate> posOffset) { builder.AddOffset(2, posOffset.Value, 0); }
  public static void AddCharacter(FlatBufferBuilder builder, int character) { builder.AddInt(3, character, 0); }
  public static void AddBoundPlayer(FlatBufferBuilder builder, Offset<Prisel.Common.PlayerInfo> boundPlayerOffset) { builder.AddOffset(4, boundPlayerOffset.Value, 0); }
  public static Offset<Monopoly.Schema.GamePlayer> EndGamePlayer(FlatBufferBuilder builder) {
    int o = builder.EndTable();
    return new Offset<Monopoly.Schema.GamePlayer>(o);
  }
};


}
