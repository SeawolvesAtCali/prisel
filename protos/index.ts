export * as monopolypb from './gen_ts/monopoly/index';
export * as priselpb from './gen_ts/prisel/index';
export * as tic_tac_toepb from './gen_ts/tic-tac-toe/index';
