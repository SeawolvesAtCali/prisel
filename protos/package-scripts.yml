scripts:
    clean: rm -rf gen_ts gen_csharp && mkdir gen_ts gen_csharp
    protoc:
        protoc --experimental_allow_proto3_optional --ts_opt force_client_none --ts_out gen_ts
        --csharp_opt base_namespace= --csharp_out gen_csharp --proto_path . $(find . -type f -name
        "*.proto")
    ts-check: tsc
    barrel: barrelsby -c barrelsby.json --delete
    rollup: rollup --config --silent
