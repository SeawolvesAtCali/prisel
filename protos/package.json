{
    "name": "@prisel/protos",
    "version": "0.1.3",
    "description": "",
    "license": "MIT",
    "author": "",
    "main": "./lib/index.cjs.js",
    "module": "./lib/index.esm.js",
    "browser": "./lib/index.umd.js",
    "types": "./dist/index.d.ts",
    "scripts": {
        "build": "nps barrel ts-check rollup",
        "test": "echo \"Error: no test specified\""
    },
    "dependencies": {
        "flatbuffers": "^2.0.3",
        "@types/long": "^4.0.0",
        "buffer": "^5.6.0",
        "long": "^4.0.0"
    },
    "devDependencies": {
        "@prisel/configs": "1.0.0",
        "barrelsby": "^2.2.0",
        "nps": "^5.9.12",
        "rollup": "^2.7.6",
        "tslib": "^2.0.0",
        "typescript": "^4.0.5"
    }
}
