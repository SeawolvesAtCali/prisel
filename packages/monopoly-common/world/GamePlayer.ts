import { monopolypb, priselpb } from '@prisel/protos';
import { list, object, serializable } from 'serializr';
import { exist } from '../exist';
import { genId } from '../genId';
import { ChanceInput } from '../types';
import { Collectible } from './Collectible';
import { GameObject } from './GameObject';
import { Id } from './Id';
import type { Property } from './Property';
import { Ref } from './ref2';
import { listRefSerializable, refSerializable } from './serializeUtil';
import type { Tile } from './Tile';

// BoundPlayer is the player.ts in @prisel/server. We don't want direct
// reference to it because player.ts is from a server side package.
interface BoundPlayer {
    getPlayerInfo(): priselpb.PlayerInfo;
}
interface Props {
    id: Id<GamePlayer>;
    player: BoundPlayer;
    pathTile: Tile;
    owning: Property[];
    money: number;
    character: number;
    rolled: boolean;
}

/**
 * GamePlayer contains the information of the player controlled character in
 * game. This is the server side presentation. When transfering game player data through protobuf, use game_player.proto.
 */
export class GamePlayer extends GameObject {
    public static TYPE = 'game_player';
    public readonly type = 'game_player';
    @serializable
    money = 0;

    @listRefSerializable
    owning: Ref<Property>[] = [];

    @refSerializable
    pathTile?: Ref<Tile>;

    @serializable
    rolled = false;

    @serializable
    character = 0;

    @serializable(list(object(Collectible)))
    collectibles: Array<Collectible> = [];

    player?: BoundPlayer;

    forcedRollPoint = 0;

    public init(props: Props) {
        this.id = props.id;
        this.pathTile = this.world.makeRef(props.pathTile);
        this.owning = (props.owning || []).map(this.world.makeRef);
        this.money = props.money;
        this.rolled = props.rolled;
        this.player = props.player;
        this.character = props.character;
        return this;
    }

    public payRent(owner: GamePlayer, property: monopolypb.PropertyInfo): monopolypb.Payment {
        this.money = this.money - property.rent;
        owner.gainMoney(property.rent);
        return {
            payer: this.id,
            payee: owner.id,
            forProperty: property,
            amount: property.rent,
        };
    }

    public gainMoney(amount: number) {
        this.money += amount;
    }

    /** Returns a random number from 1 to 6 */
    public getDiceRoll(): number {
        return Math.trunc(Math.random() * 6) + 1;
    }

    private roll(startingNode: Ref<Tile>): Tile[] {
        return this.fixedRoll(startingNode, this.getDiceRoll());
    }

    private fixedRoll(startingNode: Ref<Tile>, roll: number): Tile[] {
        return startingNode.get().genPath(roll);
    }

    public rollAndMove(fixedRoll?: number): monopolypb.Coordinate[] {
        if (exist(this.pathTile)) {
            const path = exist(fixedRoll)
                ? this.fixedRoll(this.pathTile, fixedRoll)
                : this.roll(this.pathTile);
            this.rolled = true;
            return this.move(path);
        }
        throw new Error('player not in a pathTile, cannot roll and move');
    }

    public teleport(targetTile: Tile) {
        this.pathTile = this.world.makeRef(targetTile);
    }

    public move(path: Tile[]) {
        // Currently we don't do anything different for teleport.
        // When we implement onEnter and onLeave callback for tile.
        if (path.length > 0) {
            this.pathTile = this.world.makeRef(path[path.length - 1]);
            return path.map((pathTile) => pathTile.position);
        }
        return [];
    }

    public purchaseProperty(
        property: Property,
        nextLevelPropertyInfo: monopolypb.PromptPurchaseRequest,
    ) {
        if (!nextLevelPropertyInfo.property) {
            return;
        }
        this.money = this.money - nextLevelPropertyInfo.property?.cost;
        if (property.owner?.id !== this.id) {
            this.owning.push(this.world.makeRef(property));
        }
        if (nextLevelPropertyInfo.isUpgrade) {
            property.upgrade(nextLevelPropertyInfo.property?.currentLevel);
        } else {
            property.purchasedBy(this);
        }
    }

    public getGamePlayerInfo(): monopolypb.GamePlayer {
        return {
            id: this.id,
            money: this.money,
            pos: this.pathTile?.get().position,
            character: this.character,
            boundPlayer: this.player?.getPlayerInfo(),
        };
    }

    public addCollectible(chanceInput: ChanceInput<'collectible'>) {
        this.collectibles.push({
            id: genId(),
            name: chanceInput.display.title,
            description: chanceInput.display.description,
            type: chanceInput.inputArgs.type,
        });
    }

    public hasCollectible(type: monopolypb.CollectibleExtra_CollectibleType) {
        return this.collectibles.some((collectible) => collectible.type === type);
    }

    /**
     * Pick out the collectible by removing from collection and return it.
     * @param id
     */
    public pickCollectible(id: string): Collectible | undefined {
        const collectible = this.collectibles.find((item) => item.id === id);
        this.collectibles = this.collectibles.filter((item) => item != collectible);
        return collectible;
    }

    /**
     * Pick out a collectible of type. This is used when an collectible
     * automatically become effective without player choosing, e.g.
     * get-out-of-jail card activates when going to jail
     */
    public consumeCollectible(
        type: monopolypb.CollectibleExtra_CollectibleType,
    ): Collectible | undefined {
        const collectible = this.collectibles.find((item) => item.type === type);
        this.collectibles = this.collectibles.filter((item) => item != collectible);
        return collectible;
    }
}
