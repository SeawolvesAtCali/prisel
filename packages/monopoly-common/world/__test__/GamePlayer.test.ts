import { monopolypb } from '@prisel/protos';
import { GamePlayer } from '../GamePlayer';
import { Property } from '../Property';
import { Ref } from '../ref2';
import { Tile } from '../Tile';
import { World } from '../World';

let nextId = 0;
jest.mock('../../genId', () => {
    return {
        genId: () => `mocked_id_${nextId++}`,
    };
});
beforeEach(() => {
    nextId = 0;
    // prevent the order of test affect the mocked uuid
});

describe('GamePlayer', () => {
    test('serialize', () => {
        const player = new GamePlayer();
        player.character = 1;
        player.id = '123';
        player.money = 100;
        player.owning = [Ref.forTest<Property>('property1'), Ref.forTest<Property>('property2')];
        player.addCollectible({
            display: {
                title: 'get out of jail free',
                description: 'activate when entering jail',
            },
            type: 'collectible',
            inputArgs: {
                type: monopolypb.CollectibleExtra_CollectibleType.GET_OUT_OF_JAIL_FREE,
            },
        });
        player.pathTile = Ref.forTest<Tile>('tile1');
        player.rolled = true;
        expect(player.serialize()).toMatchSnapshot();
    });

    test('deserialize', () => {
        const player = new GamePlayer();
        player.character = 1;
        player.id = '123';
        player.money = 100;
        player.owning = [Ref.forTest<Property>('property1'), Ref.forTest<Property>('property2')];
        player.pathTile = Ref.forTest<Tile>('tile1');
        player.rolled = true;
        const world = new World();
        const deserialized = GamePlayer.deserialize(player.serialize(), world);
        expect(deserialized).toBeInstanceOf(GamePlayer);
        expect(deserialized.id).toBe('123');
        expect(deserialized.money).toBe(100);
        expect(deserialized.owning).toHaveLength(2);
        expect(deserialized.owning[0].equals(Ref.forTest('property1'))).toBe(true);
        expect(deserialized.owning[1].equals(Ref.forTest('property2'))).toBe(true);
        expect(deserialized.pathTile?.equals(Ref.forTest('tile1'))).toBe(true);
        expect(deserialized.rolled).toBe(true);
        expect(deserialized.world).toBe(world);
    });
});
