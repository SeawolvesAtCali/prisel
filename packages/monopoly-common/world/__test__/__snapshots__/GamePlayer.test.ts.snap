// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePlayer serialize 1`] = `
Object {
  "character": 1,
  "collectibles": Array [
    Object {
      "description": "activate when entering jail",
      "id": "mocked_id_0",
      "name": "get out of jail free",
      "type": 1,
    },
  ],
  "id": "123",
  "money": 100,
  "owning": Array [
    "property1",
    "property2",
  ],
  "pathTile": "tile1",
  "rolled": true,
}
`;
